p2p/peer.py

```diff
@@ -790,47 +787,40 @@ def _add_peer(self,
             start_block = vm_class.dao_fork_block_number - 1
             # TODO: This can be either an `ETHPeer` or an `LESPeer`.  Will be
             # fixed once full awaitable request API is completed.
-            request = peer.request_block_headers(  # type: ignore
-                start_block,
-                max_headers=2,
-                reverse=False,
-            )
-            start = time.time()
             try:
-                while True:
-                    elapsed = int(time.time() - start)
-                    remaining_timeout = max(0, CHAIN_SPLIT_CHECK_TIMEOUT - elapsed)
-                    cmd, msg = await self.wait(
-                        peer.read_msg(), timeout=remaining_timeout)
-                    if isinstance(cmd, BaseBlockHeaders):
-                        headers = cmd.extract_headers(msg)
-                        break
-                    else:
-                        msgs.append((cmd, msg))
-                        continue
+                class MsgBuffer(PeerSubscriber):
+                    logger = logging.getLogger('p2p.peer.MsgBuffer')
+                    msg_queue_maxsize = 200
+                    subscription_msg_types = {protocol.Command}
+
+                msg_buffer = MsgBuffer()
+
+                with msg_buffer.subscribe_peer(peer):
+                    headers = await peer.handler.get_block_headers(  # type: ignore
+                        start_block,
+                        max_headers=2,
+                        reverse=False,
+                        timeout=CHAIN_SPLIT_CHECK_TIMEOUT,
+                    )
+
+                msgs = [msg_buffer.msg_queue.get_nowait()[1:] for _ in range(msg_buffer.queue_size)]
```

Comment:
Took me a bit to parse/remember that this was for capturing inbound requests for later processing. `MsgBuffer` and this is a lot of stuff that doesn't really have to do with the DAO fork. Maybe push it into `BasePeer`, which might get reused elsewhere:

```py
with peer.buffer_inbound_commands() as buffer:
  headers = await peer.handler.get_block_headers(...

msgs = buffer.get_messages()
```

Reply:
Ooh, I like this.